### PR TITLE
bytecode: Implement for loops

### PR DESCRIPTION
**Pull request by Rossiar**
Add OpStepRange and OpIterRange which push the result of evaluating 
the condition of the loop onto the stack to be evaluated by OpJumpOnFalse. 
These operations update the iteration state on the stack and also push the
current value if there is a loop variable so the bytecode can assign that value
to the variable.

Co-authored-by: joshcarp <joshcarp@users.noreply.github.com>
Co-authored-by: Cam Hutchison <camh@xdna.net>

### CODE REVIEW

File: pkg/bytecode/value.go
Comment (Rossiar) [inline pkg/bytecode/value.go:227]: We are a bit leery of this, but didn't see a better way to implement this - suggestions welcome!

File: pkg/bytecode/compiler.go
Comment (Rossiar) [inline pkg/bytecode/compiler.go:246]: I tried quite hard to unify compileIterRange and compilerStepRange, but the two functions are just different enough that any code sharing begins to look really awkward.

**Comment by camh-** [comment]
I have pushed my changes that rework this PR that addresses a few issues raised on discord:
1) It no longer re-evaluates the expressions used in the loop, as per the spec.
2) It does not create a var with an empty name when there is no loop variable (allowing nested loops with the current scopeles implementation).
3) Does the loop condition computation and comparison as part of the operations which should be faster than doing it in bytecode.
4) Ranges over strings.

This also unifies the two implementations (iter range and step range) in the compiler and also removes the map indexing by integer.

I've kept the commit that adds the `Order` field to the map value.

File: pkg/bytecode/compiler.go
Comment (camh-) [reply pkg/bytecode/compiler.go:246]: I managed to unify these largely (with a separate preamble for the two range types) by pushing a lot of the loop comparison logic into the operations.

**Review by Rossiar — COMMENTED** [review]
I can't approve as this is my PR - but lgtm

@juliaogris can you unblock us with a tick?

**Review by juliaogris — APPROVED** [review]
💚 - looks great to me, thanks @camh- for taking the time to give me personal walk through as well!

**Review by juliaogris — APPROVED** [review]
💚

**Pull request merged**
